pac4j-oauth/src/main/java/org/pac4j/oauth/client/BaseOAuth20Client.java

```diff
@@ -46,11 +48,11 @@ protected String retrieveAuthorizationUrl(final WebContext context) {
         // no request token for OAuth 2.0 -> no need to save it in the context
         final String authorizationUrl;
         // if a state parameter is required
-        if (requiresStateParameter()) {
-            String randomState = getStateParameter(context);
-            logger.debug("Random state parameter: {}", randomState);
-            context.setSessionAttribute(getName() + STATE_PARAMETER, randomState);
-            authorizationUrl = ((StateOAuth20Service) this.service).getAuthorizationUrl(randomState);
+        if (requiresStateParameter() || (stateData != null && !stateData.isEmpty())) {
```

Comment:
I don't fully agree here. Only the `requiresStateParameter` tells you if you want to use the state parameter, the value being static (your `stateData`) or dynamic (generated using the `getStateParameter()` method)